src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/AhoCorasick.cs

```diff
@@ -0,0 +1,380 @@
+
+        FastScan:
+            if (IndexOfAnyAsciiSearcher.IsVectorizationSupported && typeof(TFastScanVariant) == typeof(IndexOfAnyAsciiFastScan))
+            {
+                int remainingLength = span.Length - i;
+
+                if (remainingLength >= Vector128<ushort>.Count)
+                {
+                    // If '\0' is one of the starting chars and we're running on Ssse3 hardware, this may return false-positives.
+                    // False-positives here are okay, we'll just rule them out below. While we could flow the Ssse3AndWasmHandleZeroInNeedle
+                    // generic through, we expect such values to be rare enough that introducing more code is not worth it.
+                    int offset = IndexOfAnyAsciiSearcher.IndexOfAnyVectorized<IndexOfAnyAsciiSearcher.DontNegate, IndexOfAnyAsciiSearcher.Default>(
+                        ref Unsafe.As<char, short>(ref Unsafe.Add(ref MemoryMarshal.GetReference(span), i)),
+                        remainingLength,
+                        ref Unsafe.AsRef(_startingCharsAsciiBitmap));
+
+                    if (offset < 0)
+                    {
+                        goto Return;
+                    }
+
+                    i += offset;
+                    goto LoopWithoutRangeCheck;
+                }
+            }
+
+        Loop:
+            if ((uint)i >= (uint)span.Length)
+            {
+                goto Return;
+            }
+
+        LoopWithoutRangeCheck:
+            Debug.Assert(i < span.Length);
+            char c = TCaseSensitivity.TransformInput(Unsafe.Add(ref MemoryMarshal.GetReference(span), i));
+
+            while (true)
+            {
+                Debug.Assert((uint)nodeIndex < (uint)_nodes.Length);
+                ref AhoCorasickNode node = ref Unsafe.Add(ref nodes, (uint)nodeIndex);
+
+                if (node.TryGetChild(c, out int childIndex))
+                {
+                    nodeIndex = childIndex;
+
+                    int matchLength = Unsafe.Add(ref nodes, (uint)nodeIndex).MatchLength;
+                    if (matchLength != 0)
+                    {
+                        result = i + 1 - matchLength;
+                    }
+
+                    i++;
+                    goto Loop;
+                }
+
+                if (nodeIndex == 0)
+                {
+                    if (result >= 0)
+                    {
+                        goto Return;
+                    }
+
+                    i++;
+                    goto FastScan;
+                }
+
+                nodeIndex = node.SuffixLink;
+
+                if (nodeIndex < 0)
+                {
+                    Debug.Assert(nodeIndex == -1);
+                    Debug.Assert(result >= 0);
+                    goto Return;
+                }
+            }
+
+        Return:
+            return result;
+        }
+
+        // Mostly a copy of IndexOfAnyCore, but we may read two characters at a time in the case of surrogate pairs.
```

Comment:
Why do we need special-handling of surrogate pairs?  Can we just say we don't use this implementation if any of the strings being searched for contains surrogates, or we think that's too limiting?

And why do we need different implementations for ICU vs NLS?  With ICU as the default, can we just not have the NLS implementation and not use this strategy if it would be needed?

Reply:
Eventually versions of Windows that don't have ICU will presumably drop off our support matrix. With that plus ICU already being the default, I wouldn't think it's worth writing any new code for NLS at all.

Reply:
> Can we just say we don't use this implementation if any of the strings being searched for contains surrogates

Is it a guarantee that a surrogate pair under `IgnoreCase` will never match a non-surrogate input character?

> And why do we need different implementations for ICU vs NLS?

With ICU and Invariant, we have fast single-character or single-surrogate-pair case conversions available (`Ordinal.ToUpper`, `InvariantModeCasing.ToUpper`, `SurrogateCasing.ToUpper`).
With NLS, we have to [go through interop](https://github.com/dotnet/runtime/blob/112be1b9d055b755e5ee1b27d90c29df8cad7e6e/src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.Nls.cs#L28). I can check just how much more expensive doing that for every character would be. All `IndexOfAnyCaseInsensitiveUnicodeNls` is doing is trying to minimize that cost.

> Can we just say we don't use this implementation
> I wouldn't think it's worth writing any new code for NLS at all.

We need _some_ code (we need _a_ multi-value fallback at some point - right now this is it).
But we could go with a simpler (and NotFast:tm:) approach. This would be the most basic implementation we could use:
```c#
bool IndexOfAny(ReadOnlySpan<char> text, string[] values)
{
    for (int i = 0; i < text.Length; i++)
    {
        foreach (string value in values)
        {
            if (text.Slice(i).StartsWith(value, OrdinalIgnoreCase))
                return i;
        }
    }

    return -1;
}
```

If we're saying we don't care about NLS perf, this is fine.

Doing the same for anything that contains surrogates would be a huge perf cliff though.

Reply:
>> Can we just say we don't use this implementation if any of the strings being searched for contains surrogates

> Is it a guarantee that a surrogate pair under IgnoreCase will never match a non-surrogate input character?

If that weren't the case then presumably this would be a widespread long-standing problem?
https://github.com/dotnet/runtime/blob/76a8f4fb7b5946ec1160e93531c98548f23b8064/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L60

Reply:
I updated the `InsensitiveUnicode` path to handle NLS as well and deleted the NLS-specific helper: https://github.com/dotnet/runtime/pull/88394/commits/254f9f1a65630a5ab8e7ff294317281b2dff3650
The extra logic needed for NLS is pretty minimal now.